### PR TITLE
New pdf report

### PR DESCRIPTION
**Pull request by maxcapodi78**
Entirely based on Python

### CODE REVIEW

File: pyaedt/generic/pdf.py [hunk 1]
```diff
@@ -1,56 +1,208 @@
...
+from fpdf import FPDF
 
-class AnsysReport:
-    """Generate a pdf report."""
-
+from pyaedt import __version__
+from pyaedt.generic.constants import unit_converter
+
+
+class ReportSpec:
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:10]: If this feature is not supposed to be used in IronPython, I would suggest using dataclasses

File: pyaedt/generic/pdf.py [hunk 2]
```diff
@@ -1,56 +1,208 @@
...
+    chart_width = 16
+    font_color = [0, 0, 0]
+    font_chapter_color = [255, 0, 0]
+    font_subchapter_color = [0, 128, 0]
+    font_header_color = [0, 0, 255]
+    font_caption_color = [125, 0, 100]
+
+
+class AnsysReport(FPDF):
     def __init__(self, version="2023R1", design_name="design1", project_name="AnsysProject", tempplate_json_file=None):
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:51]:
```suggestion
    def __init__(self, version="2023R2", design_name="design1", project_name="AnsysProject", tempplate_json_file=None):
```

File: pyaedt/generic/pdf.py [hunk 3]
```diff
@@ -1,56 +1,208 @@
...
+        self._figure_idx = 1
+        self.set_top_margin(unit_converter(self.template_data.top_margin, input_units=self.template_data.units))
+        self.set_right_margin(unit_converter(self.template_data.right_margin, input_units=self.template_data.units))
+        self.set_left_margin(unit_converter(self.template_data.left_margin, input_units=self.template_data.units))
+        self.set_auto_page_break(
+            True, margin=unit_converter(self.template_data.bottom_margin, input_units=self.template_data.units)
+        )
+        self.alias_nb_pages()
+
+    def _read_template(self):
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:71]:
```suggestion
    def __read_template(self):
```

File: pyaedt/generic/pdf.py [hunk 4]
```diff
@@ -1,56 +1,208 @@
...
+        self.report_specs.ansys_version = version
+        self.report_specs.design_name = design_name
+        self.report_specs.project_name = project_name
         if tempplate_json_file:
-            self.report_specs.TemplateName = tempplate_json_file
+            self.report_specs.template_name = tempplate_json_file
+        self._read_template()
+        self._chapter_idx = 0
+        self._sub_chapter_idx = 0
+        self._figure_idx = 1
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:62]:
```suggestion
        self.__read_template()
        self.__chapter_idx = 0
        self.__sub_chapter_idx = 0
        self.__figure_idx = 1
```

File: pyaedt/generic/pdf.py [hunk 5]
```diff
@@ -1,56 +1,208 @@
...
+    def _read_template(self):
+        self.template_data = TemplateData()
+        tdata = {}
+        with open(self.report_specs.template_name, "r") as f:
+            tdata = json.load(f)
+        for k, v in tdata.items():
+            if k in self.template_data.__dict__:
+                self.template_data.__dict__[k] = v
+
+    def _add_cover_page(self):
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:80]:
```suggestion
    def __add_cover_page(self):
```

File: pyaedt/generic/pdf.py [hunk 6]
```diff
@@ -1,56 +1,208 @@
...
         """
-        self.report = CreatePdfReport(os.path.join(pyaedt_path, "dlls", "PDFReport"))
-        self.report.Specs = self.report_specs
-        if add_header:
-            self.report.AddAnsysHeader(-1)
-        if add_first_page:
-            self.report.AddFirstPage()
-        return self.report
+        if add_cover_page:
+            self._add_cover_page()
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:202]:
```suggestion
            self.__add_cover_page()
```

File: pyaedt/generic/pdf.py [hunk 7]
```diff
@@ -1,56 +1,208 @@
...
+        self.set_x(self.w - self.r_margin - 33)
+        self.image(
+            self.report_specs.logo_name,
+            self.x,
+            self.y,
+            unit_converter(self.template_data.header_image_width, input_units=self.template_data.units),
+        )
+        self.set_x(self.l_margin)
+        self.set_y(self.t_margin)
+        self.line(x1=self.l_margin, y1=self.t_margin - 7, x2=self.w - self.r_margin, y2=self.t_margin - 7)
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:174]:
```suggestion
    def header(self):
        from datetime import date

        def add_field(field_name, field_value):
            self.set_font(self.template_data.font.lower(), size=self.template_data.header_font_size)
            self.cell(
                0,
                3,
                field_name,
                new_x="LMARGIN",
                new_y="NEXT",
                align="L",
            )
            self.set_x(line_x)
            self.cell(
                0,
                3,
                field_value,
                new_x="LMARGIN",
                new_y="NEXT",
                align="L",
            )
        
        # Logo
        self.set_y(15)
        line_x = self.x
        line_y = self.y
        delta = (self.w - self.r_margin - self.l_margin) / 5 - 10
        self.set_text_color(*self.template_data.font_header_color)

        add_field("Project Name", self.report_specs.project_name)
        self.set_y(line_y)
        line_x += delta
        self.set_x(line_x)
        add_field("Design Name", self.report_specs.design_name)

        self.set_y(line_y)
        line_x += delta
        self.set_x(line_x)
        add_field("Ansys Version", self.report_specs.ansys_version)
        self.set_y(line_y)
        line_x += delta
        self.set_x(line_x)
        add_field("PyAEDT Version", self.report_specs.pyaedt_version)
        self.set_y(line_y)
        line_x += delta
        self.set_x(line_x)

        add_field("Date", str(date.today()))
        self.set_y(line_y)
        line_x += delta
        self.set_x(line_x)
        add_field("Revision", self.report_specs.revision)
        self.set_y(10)
        line_x += delta
        self.set_x(self.w - self.r_margin - 33)
        self.image(
            self.report_specs.logo_name,
            self.x,
            self.y,
            unit_converter(self.template_data.header_image_width, input_units=self.template_data.units),
        )
        self.set_x(self.l_margin)
        self.set_y(self.t_margin)
        self.line(x1=self.l_margin, y1=self.t_margin - 7, x2=self.w - self.r_margin, y2=self.t_margin - 7)
```

File: pyaedt/generic/pdf.py [hunk 8]
```diff
@@ -104,53 +256,91 @@ def aedt_version(self):
...
         Parameters
         ----------
         chapter_name : str
             Chapter name.
-        section_id : int, optional
-            Section on which add the chapter. Default is 0 which use current section.
         """
-        self.report.AddChapter(chapter_name, section_id)
+        self._chapter_idx += 1
+        self._sub_chapter_idx = 0
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:294]:
```suggestion
        self.__chapter_idx += 1
        self.__sub_chapter_idx = 0
```

File: pyaedt/generic/pdf.py [hunk 9]
```diff
@@ -104,53 +256,91 @@ def aedt_version(self):
...
+        txt = f"{self._chapter_idx} {chapter_name}"
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:295]:
```suggestion
        txt = f"{self.__chapter_idx} {chapter_name}"
```

Comment (SMoraisAnsys) [reply pyaedt/generic/pdf.py:295]: @maxcapodi78 (warning) here you are using f-strings and I don't think that they are available in IronPython (or at least I coudn't find them), if this feature isn't intended to be used in IronPython, I would advise you follow my first comment an use dataclasses for the top class of this file.

File: pyaedt/generic/pdf.py [hunk 10]
```diff
@@ -104,53 +256,91 @@ def aedt_version(self):
...
         Parameters
         ----------
         chapter_name : str
             Chapter name.
-        section_id : int, optional
-            Section on which add the subchapter. Default is 0 which use current section.
         """
-        return self.report.AddSubChapter(chapter_name, section_id)
+        self._sub_chapter_idx += 1
+        txt = f"     {self._chapter_idx}.{self._sub_chapter_idx} {chapter_name}"
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:320]:
```suggestion
        self.__sub_chapter_idx += 1
        txt = f"     {self.__chapter_idx}.{self.__sub_chapter_idx} {chapter_name}"
```

Comment (SMoraisAnsys) [reply pyaedt/generic/pdf.py:320]: cf previous comment

File: pyaedt/generic/pdf.py [hunk 11]
```diff
@@ -160,55 +350,65 @@ def add_image(self, path, caption="", width=10, section_id=0):
...
-        return self.report.AddImageWithCaption(path, caption, width, section_id)
+        if width == 0:
+            width = self.epw
 
-    def add_caption(self, content, section_id=0):
+        self.image(path, h=height, w=width, x=self.epw / 2 - width / 2 + self.l_margin)
+        if caption:
+            caption = f"Figure {self._figure_idx}. {caption}"
+            self.add_caption(caption)
+            self._figure_idx += 1
```
Comment (SMoraisAnsys) [inline pyaedt/generic/pdf.py:364]:
```suggestion
            caption = f"Figure {self.__figure_idx}. {caption}"
            self.add_caption(caption)
            self.__figure_idx += 1
```

**Review by SMoraisAnsys — CHANGES_REQUESTED** [review]
Good job with this refactoring ! I do have a warning to raise and a couple of minor changes.

Warning: you use f-strings (which is nice) but I don't know if this feature should be used by our IronPython users. If that's the case, you should this and use str `format` to be compatible. If that's not the case, I would recommend using `dataclasses` for the classes at the beginning of the file pdf.py. That would make total sense here and you could just use `MyDataClass(**kwargs)`.

Minor changes:
- use private method and attribute instead of protected (`__` vs `_`) except for self._outline which comes from `FPDF`;
- move import at the begining of a function

File: pyaedt/generic/pdf.py (hunk 9 above)
Comment (maxcapodi78) [reply pyaedt/generic/pdf.py:295]: this has not to be compatible with IPY because fpdf is not supported by IPY

File: pyaedt/generic/pdf.py (hunk 1 above)
Comment (maxcapodi78) [reply pyaedt/generic/pdf.py:10]: it's a good idea

**Review by Samuelopez-ansys — APPROVED** [review]
Really great! It is worth adding it somewhere in the documentation for better feature visibility.

**Review by SMoraisAnsys — APPROVED** [review]

**Pull request merged**